[batch] Use the default region when a user doesn't specify a region

jigold opened the pull request:
## Change Description

BREAKING CHANGE: This changes the behavior of what the default regions are set to from Any Region to the Batch Hosted Region.

This PR is intended to make the default behavior of the hailtop.batch client use only the default region if no regions are specified. We need to make this change because this is an obvious place for cost savings as knowing to set the regions may not be conveyed to new people or others may forget when getting a new laptop.

There are other ways to implement this change by using environment variables instead of "Global Config" objects. I'm not particular on how it's actually implemented.

## Security Assessment

- This change has a low security impact

### Impact Description

Added a new API endpoint to the Batch front end. It should only be accessible to Authorized Users.

(Reviewers: please confirm the security impact before approving)

jigold commented on hail/python/test/hailtop/batch/test_batch_service_backend.py:
```diff
@@ -507,6 +507,21 @@ def test_specify_job_region(service_backend: ServiceBackend):
+def test_use_default_region_when_not_specifying_any_regions(service_backend: ServiceBackend):
+    b = batch(service_backend)
+    j = b.new_job('region')
+    default_region = service_backend.default_region()
```
It would probably be good to assert this is "us-central1" somehow just to make absolutely sure it has the behavior we want either in the test or double checking the global config secret.

patrick-schultz replied:
I don't think we want to hardcode config in the tests, for the sake of people running forks. But I did just confirm that the default is "us-central1" in the global config secret.

jigold commented:
@samcbryant @patrick-schultz @ehigham 

Is it possible to start making some progress on this change whether it's an initial review of the code or a discussion of the change? Thank you in advance!

patrick-schultz commented:
@cjllanwarne If you have a few minutes today, could you give this a look?

patrick-schultz commented on hail/python/hailtop/batch/backend.py:
```diff
@@ -618,8 +634,11 @@ def __init__(
                 assert isinstance(regions_from_conf, str)
                 regions = regions_from_conf.split(',')
+            else:
+                regions = [ServiceBackend.default_region()]
```
Need to update the docstring for the `regions` parameter to reflect the new default behavior.

patrick-schultz reviewed (changes requested):
Thanks Jackie, this is another good ux improvement. Everything looks good to me, just need to update the docstring to reflect the change.

patrick-schultz reviewed (approved):
Thanks Jackie!

jigold commented:
I pushed a fix. Not sure if you need to put the new SHA into CI

jigold commented:
Sorry, overlooked that one. Hopefully this passes now.

patrick-schultz commented on hail/python/test/hailtop/batch/test_batch_service_backend.py:
```diff
@@ -507,6 +507,21 @@ def test_specify_job_region(service_backend: ServiceBackend):
+    b = batch(service_backend)
+    j = b.new_job('region')
+    default_region = service_backend.default_region()
+    j.regions()
```
This needs an argument. What was the intention? Don't you want to test when the region isn't set explicitly?

jigold replied:
I wanted to make sure it was set to None in case the default was set anywhere in the test suite. I'm just passing `None`.

patrick-schultz commented on hail/python/test/hailtop/batch/test_batch_service_backend.py:
```diff
@@ -507,6 +507,21 @@ def test_specify_job_region(service_backend: ServiceBackend):
+    default_region = service_backend.default_region()
+    j.regions()
+    j.command('true')
+    assert b._regions == [default_region], str(b._regions)
```
The attribute on Batch is `_default_regions`

patrick-schultz reviewed (changes requested):
Okay, this looks like the last issue

jigold commented:
It failed :( Can I get the log? Thanks!

chrisvittal commented:
This seems to be a config issue:
```
___________ test_use_default_region_when_not_specifying_any_regions ____________

service_backend = <hailtop.batch.backend.ServiceBackend object at 0x74934aa6dbe0>

    def test_use_default_region_when_not_specifying_any_regions(service_backend: ServiceBackend):
>       assert service_backend.regions is None
E       AssertionError: assert ['us-central1'] is None
E        +  where ['us-central1'] = <hailtop.batch.backend.ServiceBackend object at 0x74934aa6dbe0>.regions
```

chrisvittal commented:
The other errors seem transient

jigold commented:
Hopefully I fixed it now. I just decided not to use the fixture for the session and explicitly set the regions to None.

chrisvittal commented:
I expanded the traceback for this failing test:
```
_______________ test_cant_submit_to_default_with_other_ns_creds ________________

client = <hailtop.batch_client.client.BatchClient object at 0x76b0d7ff15e0>
remote_tmpdir = 'gs://hail-test-ezlis/test_batch/gkca1adm9xsn'

    def test_cant_submit_to_default_with_other_ns_creds(client: BatchClient, remote_tmpdir: str):
        DOMAIN = os.environ['HAIL_PRODUCTION_DOMAIN']
        NAMESPACE = os.environ['HAIL_DEFAULT_NAMESPACE']
    
        script = f"""import hailtop.batch as hb
    backend = hb.ServiceBackend("test", remote_tmpdir="{remote_tmpdir}")
    b = hb.Batch(backend=backend)
    j = b.new_bash_job()
    j.command("echo hi")
    b.run()
    backend.close()
    """
    
        b = create_batch(client)
        j = b.create_job(
            HAIL_GENETICS_HAILTOP_IMAGE,
            [
                '/bin/bash',
                '-c',
                f"""
    python3 -c \'{script}\'""",
            ],
            env={
                'HAIL_DOMAIN': DOMAIN,
                'HAIL_DEFAULT_NAMESPACE': 'default',
                'HAIL_LOCATION': 'external',
                'HAIL_BASE_PATH': '',
            },
        )
        b.submit()
        status = j.wait()
        if NAMESPACE == 'default':
            assert status['state'] == 'Success', str((status, b.debug_info()))
        else:
            assert status['state'] == 'Failed', str((status, b.debug_info()))
>           assert 'Unauthorized' in j.log()['main'], (str(j.log()['main']), status)
E           AssertionError: ("/usr/local/lib/python3.9/dist-packages/hailtop/batch/backend.py:586: UserWarning: Use of deprecated positional argum...7e76e9e3760>
E             ", {'always_run': False, 'batch_id': 164, 'billing_project': 'test', 'cost': 8.717416793838885e-05, ...})
E           assert 'Unauthorized' in '/usr/local/lib/python3.9/dist-packages/hailtop/batch/backend.py:586: UserWarning: Use of deprecated positional argument \'billing_project\' in ServiceBackend(). Specify \'billing_project\' as a keyword argument instead.
  warnings.warn(
Traceback (most recent call last):
  File "<string>", line 2, in <module>
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch/backend.py", line 642, in __init__
    regions = [ServiceBackend.default_region()]
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch/backend.py", line 567, in default_region
    return dummy_client.default_region()
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch_client/client.py", line 456, in default_region
    return async_to_blocking(self._async_client.default_region())
  File "/usr/local/lib/python3.9/dist-packages/hailtop/utils/utils.py", line 186, in async_to_blocking
    raise exc
  File "/usr/local/lib/python3.9/dist-packages/hailtop/utils/utils.py", line 181, in async_to_blocking
    return loop.run_until_complete(task)
  File "/usr/local/lib/python3.9/dist-packages/nest_asyncio.py", line 98, in run_until_complete
    return f.result()
  File "/usr/lib/python3.9/asyncio/futures.py", line 201, in result
    raise self._exception
  File "/usr/lib/python3.9/asyncio/tasks.py", line 256, in __step
    result = coro.send(None)
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch_client/aioclient.py", line 1441, in default_region
    resp = await self._get(\'/api/v1alpha/default_region\')
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch_client/aioclient.py", line 1322, in _get
    await self._session.get(self.url + path, params=params, headers=self._headers)
  File "/usr/local/lib/python3.9/dist-packages/hailtop/aiocloud/common/session.py", line 25, in get
    return await self.request(\'GET\', url, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/hailtop/aiocloud/common/session.py", line 105, in request
    return await retry_transient_errors(self._request_with_valid_authn, method, url, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/hailtop/utils/utils.py", line 792, in retry_transient_errors
    return await retry_transient_errors_with_debug_string(\'\', 0, f, *args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/hailtop/utils/utils.py", line 809, in retry_transient_errors_with_debug_string
    return await f(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/hailtop/aiocloud/common/session.py", line 117, in _request_with_valid_authn
    return await self._http_session.request(method, url, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/hailtop/httpx.py", line 148, in request_and_raise_for_status
    raise ClientResponseError(
hailtop.httpx.ClientResponseError: 404, message=\'Not Found\', url=URL(\'https://batch.hail.is/api/v1alpha/default_region\') body=\'404: Not Found\'
Exception ignored in: <function Backend.__del__ at 0x77e76ea3c280>
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch/backend.py", line 123, in __del__
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch/backend.py", line 115, in close
  File "/usr/local/lib/python3.9/dist-packages/hailtop/utils/utils.py", line 186, in async_to_blocking
  File "/usr/local/lib/python3.9/dist-packages/hailtop/utils/utils.py", line 181, in async_to_blocking
  File "/usr/local/lib/python3.9/dist-packages/nest_asyncio.py", line 98, in run_until_complete
  File "/usr/lib/python3.9/asyncio/futures.py", line 201, in result
  File "/usr/lib/python3.9/asyncio/tasks.py", line 256, in __step
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch/backend.py", line 119, in async_close
  File "/usr/local/lib/python3.9/dist-packages/hailtop/batch/backend.py", line 665, in _async_close
AttributeError: \'ServiceBackend\' object has no attribute \'_ServiceBackend__batch_client\'
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x77e76e9e3a00>
Unclosed connector
connections: [\'deque([(<aiohttp.client_proto.ResponseHandler object at 0x77e76e9ddee0>, 208.987484927)])\']
connector: <aiohttp.connector.TCPConnector object at 0x77e76e9e3910>
Unclosed client session
client_session: <aiohttp.client.ClientSession object at 0x77e76e9e3850>
Unclosed connector
connections: [\'deque([(<aiohttp.client_proto.ResponseHandler object at 0x77e76e9dda00>, 208.952670645)])\']
connector: <aiohttp.connector.TCPConnector object at 0x77e76e9e3760>
'
```

chrisvittal commented:
this is the other failure:
```
___________ test_use_default_region_when_not_specifying_any_regions ____________

    def test_use_default_region_when_not_specifying_any_regions():
        service_backend = ServiceBackend()
        default_region = service_backend.default_region()
    
        try:
            assert service_backend.regions == [default_region], str(service_backend.regions)
    
            # test default works when service backend `regions` has no default set
            service_backend.regions = None
    
            b = batch(service_backend, default_regions=None)
            j = b.new_job('region')
            j.command('true')
    
>           assert b._default_regions == [default_region], str(b._default_regions)
E           AssertionError: None
E           assert None == ['us-central1']
E            +  where None = <hailtop.batch.batch.Batch object at 0x7b8714a91ee0>._default_regions
```

patrick-schultz commented:
Ah, thanks Chris. I was searching for "failed" in the output, and that jumped over the stacktraces 🤦

jigold commented:
I think I fixed the first bug. The second requires that we split this PR into two and make the server-side change "adding the default_region" API route and then follow up with the client side changes. Sound good?

jigold commented:
I hope this should all pass now 🤞

chrisvittal reviewed (approved):
Excellent, and this no longer touches the backend outside of making use of the new API.
